CharacTER: MT metric

BramVanroy opened the pull request:
Add the CharacTER MT evaluation metric that was introduced in [CharacTer: Translation Edit Rate on Character Level](https://aclanthology.org/W16-2342).

Specifically, this implementation uses [the repackaged version ](https://github.com/BramVanroy/CharacTER) of [the original](https://github.com/rwth-i6/CharacTER) for usability reasons.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/evaluate/pr_286). All of your documentation changes will be reflected on that endpoint.

BramVanroy commented:
Experiencing some isort issues like `WARNING: Unable to parse file tests due to [Errno 13] Permission denied: 'C:\\Python\\projects\\evaluate\\tests'`. Will investigate later.

mathemakitten commented:
Thanks @BramVanroy, this is cool! Will add to my list to review mid-next week :)

BramVanroy commented:
@mathemakitten I fixed the isort issues.

mathemakitten commented on metrics/character/README.md:
````diff
@@ -0,0 +1,94 @@
+results = character.compute(references=refs, predictions=preds)
+
+```
+### Inputs
+- **predictions**: a single prediction or a list of predictions to score. Each prediction should be a string with
+     tokens separated by spaces.
+- **references**: a single reference or a list of reference for each prediction. Each reference should be a string with
+     tokens separated by spaces.
````
Small thing: I think this currently fails on single examples since the CER package tries to calculate stdev, but that requires at least two data points (see below for example). Do you know if there's a way to pass a flag down to `cer` to turn this off for single examples?

```
p = ["this week the saudis denied information published in the new york times"]
r = ["saudi arabia denied this week information published in the american new york times"]
results = character.compute(references=r, predictions=p)
```

<img width="1028" alt="Screen Shot 2022-09-13 at 11 56 21 AM" src="https://user-images.githubusercontent.com/31600291/189986779-31b99a4f-f55d-48c0-b7e3-39f1e3ae60e6.png">

Passing in strings directly also triggers a mismatch error (`Mismatch in the number of predictions (71) and references (82)`).

BramVanroy replied:
Ah, good call. I'll update `cer` one of these days to make sure that does not give any issues.

mathemakitten commented on metrics/character/README.md:
```diff
@@ -0,0 +1,94 @@
+
+
+### Output Values
+- **count**: how many parallel sentences were processed
```
```suggestion
- **count**: number of sentence pairs processed
```

mathemakitten commented on metrics/character/character.py:
```diff
@@ -0,0 +1,123 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""An implementation for calculating CharacTER, a character-based TER variant, useful for machine translation tasks."""
```
```suggestion
"""CharacTER metric, a character-based TER variant, for machine translation."""
```

mathemakitten commented on metrics/character/character.py:
```diff
@@ -0,0 +1,123 @@
+"""
+
+_DESCRIPTION = """\
+CharacTer is a novel character level metric inspired by the commonly applied translation edit rate (TER). It is
```
See suggested changes re: this blurb above :)

mathemakitten commented on metrics/character/character.py:
```diff
@@ -0,0 +1,123 @@
+
+@evaluate.utils.file_utils.add_start_docstrings(_DESCRIPTION, _KWARGS_DESCRIPTION)
+class Character(evaluate.Metric):
+    """CharacTer is a novel character level metric inspired by the commonly applied translation edit rate (TER)."""
```
```suggestion
    """CharacTer is a character-level metric inspired by the commonly applied translation edit rate (TER)."""
```

mathemakitten commented on metrics/character/character.py:
```diff
@@ -0,0 +1,123 @@
+class Character(evaluate.Metric):
+    """CharacTer is a novel character level metric inspired by the commonly applied translation edit rate (TER)."""
+
+    def _info(self):
+        return evaluate.MetricInfo(
+            # This is the description that will appear on the modules page.
+            module_type="metric",
+            description=_DESCRIPTION,
+            citation=_CITATION,
+            inputs_description=_KWARGS_DESCRIPTION,
+            # This defines the format of each prediction and reference
+            features=[
+                datasets.Features(
+                    {"predictions": Value("string", id="prediction"), "references": Value("string", id="reference")}
+                ),
+                datasets.Features(
+                    {
+                        "predictions": Sequence(Value("string", id="prediction"), id="predictions"),
+                        "references": Sequence(Value("string", id="reference"), id="references"),
+                    }
+                ),
+            ],
+            # Homepage of the module for documentation
+            homepage="https://github.com/bramvanroy/CharacTER",
+            # Additional links to the codebase or references
+            codebase_urls=["https://github.com/bramvanroy/CharacTER", "https://github.com/rwth-i6/CharacTER"],
+        )
```
```suggestion
    def _info(self):
        return evaluate.MetricInfo(
            module_type="metric",
            description=_DESCRIPTION,
            citation=_CITATION,
            inputs_description=_KWARGS_DESCRIPTION,
            features=[
                datasets.Features(
                    {"predictions": Value("string", id="prediction"), "references": Value("string", id="reference")}
                ),
                datasets.Features(
                    {
                        "predictions": Sequence(Value("string", id="prediction"), id="predictions"),
                        "references": Sequence(Value("string", id="reference"), id="references"),
                    }
                ),
            ],
            homepage="https://github.com/bramvanroy/CharacTER",
            codebase_urls=["https://github.com/bramvanroy/CharacTER", "https://github.com/rwth-i6/CharacTER"],
        )
```

mathemakitten replied:
These are mostly cosmetic suggestions

mathemakitten commented on metrics/character/README.md:
```diff
@@ -0,0 +1,94 @@
+         "this is in fact an estimate"]
+refs = ["saudi arabia denied this week information published in the american new york times",
+        "this is actually an estimate"]
+results = character.compute(references=refs, predictions=preds)
```
Could we also add an example for the single-pair example here? :)

mathemakitten commented on metrics/character/character.py:
```diff
@@ -0,0 +1,123 @@
+Args:
+    predictions: a single prediction or a list of predictions to score. Each prediction should be a string with
+     tokens separated by spaces.
+    references: a single reference or a list of reference for each prediction. Each reference should be a string with
```
```suggestion
    references: a single reference or a list of references for each prediction. Each reference should be a string with
```

mathemakitten commented on metrics/character/character.py:
```diff
@@ -0,0 +1,123 @@
+    >>> results = character.compute(references=refs, predictions=preds)
+    >>> print(results)
+    {
+        'count': 2,
```
Question: is there a way to return the per-example / unaggregated score for each example as well?

mathemakitten commented on metrics/character/tests.py:
```diff
@@ -0,0 +1,35 @@
+from statistics import mean, median, stdev
```
@lvwerra, do you have a suggestion for where tests like these should go? We don't currently keep metric-specific tests in the metric folders, but these are handy to have around. (@BramVanroy FYI Leandro is on vacation until next week!)

BramVanroy replied:
If I remember correctly, the `tests.py` is created automatically when using `evaluate-cli create`.

lvwerra replied:
Hi @BramVanroy and @mathemakitten, the `tests.py` is WIP and currently does nothing. The idea was originally that we could setup a list of example inputs/outputs in that file that we could then use to test the metric as well as use them as examples in the Gradio app. 

The metrics are currently tested with doctest. That means the examples in the metric's docstring are executed in the tests. So if you think there are important edge cases to be tested you could add them there.

mathemakitten reviewed (commented):
Hi @BramVanroy, thanks so much for putting this together! I added some comments and questions, please have a look and let me know if I can clarify anything 🤗

BramVanroy commented:
> Hi @BramVanroy, thanks so much for putting this together! I added some comments and questions, please have a look and let me know if I can clarify anything 🤗

Thanks for having such a detailed look @mathemakitten! I incorporated your suggestions. I am encountering an issue again with the required format. I have experienced this behavior and I am not sure how to solve this.

It seems that when I pass a single sentence (not a list) the string is used as a sequence of letters, with the following errors:

```
ValueError: Mismatch in the number of predictions (71) and references (82)
```

You can trigger this error by running `python -m pytest .\metrics\character\tests.py`.

When running `compute`, what kind of format is expected? I tried adding a batch dimension, but that will pass the list to `compute`, which in turn will trigger `cer.calculate_cer_corpus` instead of `calculate_cer`. So it seems I can never "just" pass a string? Please advise me on that.

lvwerra commented:
> When running compute, what kind of format is expected? I tried adding a batch dimension, but that will pass the list to compute, which in turn will trigger cer.calculate_cer_corpus instead of calculate_cer. So it seems I can never "just" pass a string? Please advise me on that.

So compute **always** expects a list of samples. So if you want to use a feature consists of `Value("string")` you should pass the following:

```python
metric.compute(predictions=["my example"], references=["your example"])
```

If you only want to pass a single example you can use `add`:
```python
metric.add(predictions="my example", references="your example")
```

Finally, if your features themselves are lists (e.g. tokenized strings), you should pass them as follows:
```python
metric.compute(predictions=[["my", "example"]], references=[["your", "example"]])
```

Does that make sense? In your case it indeed seems that a string is interpreted as a list and that might be causing the mismatch so I am guessing you passed the `add` format above to `compute`. If not, could you share an example?

lvwerra reviewed (commented):
Looks good. I wonder if it would make sense to extend it to the case where you have multiple references? E.g. take the minimum in that case, which corresponds to the score to the most similar reference? This would make it easy to combine it with all the other text metrics. What do you think?

BramVanroy commented:
> Looks good. I wonder if it would make sense to extend it to the case where you have multiple references? E.g. take the minimum in that case, which corresponds to the score to the most similar reference? This would make it easy to combine it with all the other text metrics. What do you think?

Sure, I can do that. Should it also return the most similar reference in that case, or just the score?

BramVanroy commented:
I implemented my suggestion above. Please let me know if that is inline with what you expected @lvwerra, so I can update the PR in charcut in a similar way.

lvwerra reviewed (approved):
That's perfect, thanks for adding these features! 🚀

lvwerra commented:
Looks like `Literal` was added only in Python 3.8 (we are still supporting 3.7). Could you fix this? If the CI is green we can merge :)

BramVanroy commented:
Done!

lvwerra commented:
I saw that there are some suggestions from @mathemakitten left, would you mind adding them?

BramVanroy commented:
> I saw that there are some suggestions from @mathemakitten left, would you mind adding them?

Done.
